### PR TITLE
feat: [LW-7837] add warning banner for retiring, retired and saturated stake pools

### PR DESCRIPTION
**Pull request by greatertomi**
# Checklist

- [x] JIRA - LW-7837
- [ ] Proper tests implemented
- [x] Screenshots added.

---

## Proposed solution

Add a warning banner when any pool in the user portfolio is saturated, retiring or retired.

## Screenshots

Screenshot 2023-10-23 at 20 29 43

---
# Allure report
`allure-report-publisher` generated test report!

**smokeTests**: ❌ [test report](https://lace-qa:8PP5wBaDV6UoXj@dq4ajm5i5q7bz.cloudfront.net/linux/chrome/2715/6653412193/index.html) for [42341835](https://github.com/input-output-hk/lace/pull/659/commits/42341835db317cdee502706c0e0deab865bd0aaa)
|       | passed | failed | skipped | flaky | total | result |
|-------|--------|--------|---------|-------|-------|--------|
| Total | 31     | 1      | 0       | 0     | 32    | ❌     |

### CODE REVIEW

File: packages/staking/src/features/overview/StakingNotificationBanner/StakingNotificationBanner.tsx [hunk 1]
```diff
@@ -50,6 +51,17 @@ export const StakingNotificationBanner = ({
...
       );
+    case 'poolSaturatedOrRetired':
+      return (
+        <Banner
+          popupView={popupView}
+          withIcon
+          customIcon={<ExclamationIcon />}
+          message={t('overview.banners.saturatedOrRetiredPool.title')}
+          description={t('overview.banners.saturatedOrRetiredPool.message')}
+          onBannerClick={onPortfolioDriftedNotificationClick}
```
Comment (refi93) [inline packages/staking/src/features/overview/StakingNotificationBanner/StakingNotificationBanner.tsx:62]: the `onPortfolioDriftedNotificationClick` name no longer makes sense as it's being used also for the case of saturated/retired pools which is not related to portfolio drift. To tackle this, I'd consider generalizing the name of the prop to `clickableBannerAction`

Comment (refi93) [reply packages/staking/src/features/overview/StakingNotificationBanner/StakingNotificationBanner.tsx:62]: alternatively, we can just encapsulate this logic into `StakingNotificationBanner` and bind it to `isPopupView` flag (i.e. open management drawer if expanded view, otherwise open expanded view) - I think that would indeed be cleaner

File: packages/staking/src/features/overview/StakingNotificationBanner/getCurrentStakingNotification.ts [hunk 2]
```diff
@@ -22,5 +22,9 @@ export const getCurrentStakingNotification = ({
     return 'portfolioDrifted';
   }
 
+  if (hasSaturatedOrRetiredPools(currentPortfolio)) {
```
Comment (refi93) [inline packages/staking/src/features/overview/StakingNotificationBanner/getCurrentStakingNotification.ts:25]: I don't see in the task the priority of the "saturated or retired" notification vs the "drift" one and I can see arguments in favor of either ordering. I'd suggest clarifying the precence of notification in such a case with Fernando, if you already didn't

File: packages/staking/src/features/overview/helpers/hasSaturatedOrRetiredPools.ts [hunk 3]
```diff
@@ -0,0 +1,11 @@
+import { CurrentPortfolioStakePool } from '../../store';
+
+export const SATURATION_UPPER_BOUND = 100;
```
Comment (refi93) [inline packages/staking/src/features/overview/helpers/hasSaturatedOrRetiredPools.ts:3]: I'd consider extracting this constant which is already used in the staking package to a `constants.ts` file within the `helpers` folder

File: packages/staking/src/assets/icons/triangle-exclamation.component.svg
Comment (refi93) [inline packages/staking/src/assets/icons/triangle-exclamation.component.svg:1]: to prevent further duplication/inconsistencies of assets I'd consider moving this (together with other icons in this folder, for the sake of consistency) to `packages/ui/assets/icons` folder from where we could import them as we already do e.g. for `chevron-up` icon: https://github.com/input-output-hk/lace/blob/0ce732940b0aea274e9036c491863c21e409b607/packages/staking/src/features/drawer/preferences/PoolDetailsCard/PoolDetailsCard.tsx#L3C1-L3C1

**Review by refi93 — CHANGES_REQUESTED** [review]
see comments

File: packages/staking/src/features/overview/StakingNotificationBanner/getCurrentStakingNotification.ts (hunk 2 above)
Comment (greatertomi) [reply packages/staking/src/features/overview/StakingNotificationBanner/getCurrentStakingNotification.ts]: I asked Fernando and he said we should show both banners. 8a1ba0d does that.

File: packages/staking/src/features/overview/helpers/hasSaturatedOrRetiredPools.ts (hunk 3 above)
Comment (greatertomi) [reply packages/staking/src/features/overview/helpers/hasSaturatedOrRetiredPools.ts:3]: Fixed on 82593bd

**Review by mrcnk — APPROVED** [review]

File: packages/staking/src/features/overview/Overview.tsx [hunk 4]
```diff
@@ -111,6 +118,17 @@ export const Overview = () => {
           />
         </Box>
       )}
+      {isPoolRetiredOrSaturated && (
+        <Box mb="$40">
+          <Banner
```
Comment (refi93) [inline packages/staking/src/features/overview/Overview.tsx:123]: I'd consider keeping this Banner within the `StakingNotificationBanner` component to keep the notification logic in a single place, otherwise it may become hard to follow what notifications are shown when which may result in undesired combinations of the notifications being shown

I'd suggest refactoring `getCurrentStakingNotification()` to return an array of notifications (most of the time it would be empty or single item, with the exception of saturated/retired pool and drift at the same time) and then this Banner component should be movable into `StakingNotificationBanner` component (which I'd also refactor to take an array of notifications and rename it to `StakingNotificationBanners`

What do you think?

**Review by refi93 — CHANGES_REQUESTED** [review]
see comment

File: packages/staking/src/features/overview/Overview.tsx (hunk 4 above)
Comment (greatertomi) [reply packages/staking/src/features/overview/Overview.tsx]: I agree. Created an array of notifications here: 97d92bb

File: packages/staking/src/features/overview/StakingNotificationBanner/StakingNotificationBanner.tsx (hunk 1 above)
Comment (refi93) [reply packages/staking/src/features/overview/StakingNotificationBanner/StakingNotificationBanner.tsx]: agreed offline to go with the first suggestion, but naming the flag `onClickableBannerClick` which is simpler while it should make clear enough to the reader that not all banners are clickable

File: packages/staking/src/features/overview/StakingNotificationBanner/StakingNotificationBanners.tsx [hunk 5]
```diff
@@ -0,0 +1,70 @@
+import { Banner } from '@lace/common';
+import ExclamationIcon from '@lace/ui/dist/assets/icons/warning-icon-triangle.component.svg';
+import { useTranslation } from 'react-i18next';
+import BellIcon from '../../../assets/icons/bell-icon.component.svg';
+import InfoIcon from '../../../assets/icons/info-icon.svg';
+import * as styles from './StakingNotificationBanner.css';
+import { StakingNotificationType } from './types';
+
+type StakingNotificationBannerProps = {
```
Comment (refi93) [inline packages/staking/src/features/overview/StakingNotificationBanner/StakingNotificationBanners.tsx:9]:
```suggestion
type StakingNotificationBannersProps = {
```

**Review by refi93 — APPROVED** [review]

**Review by oldGreg5 — APPROVED** [review]
